### ISSUE TITLE
Support for PEFT in GPRO

### ISSUE THREAD

**Issue by sravan500** [NVIDIA-NeMo/RL#833]
Hi Team,

I understand that Nemo RL GRPO supports only full finetuning now,Can we expect support for adapter finetuning feature.

**Comment by sravan500** [comment]
Hi Team,

Yes we want to train models using Lora.
Hardware we want to use H100.

**Comment by sravan500** [comment]
Can anyone please respond.
Is adding LoRa in the roadmap of NeMo RL?

**Comment by terrykong** [comment]
Hi @sravan500 . LoRa is on the roadmap, but we currently don't have an ETA. For context, we are currently transitioning our pure pytorch backend ("dtensor") to use https://github.com/NVIDIA-NeMo/Automodel. Once we do, we should be able to enable this

**Comment by afennelly-mitre** [comment]
> We will get this for v0.5, which is targeting Oct release.

any update on this, given that we are still waiting for v0.4 release?

**Comment by snowmanwwg** [comment]
Update - basica implementation is done in the PyTorch/Automodel backend. Doing some final accuracy debugging. Hope to release soon.

**Comment by sravan500** [comment]
Thanks for the update.

**Comment by snowmanwwg** [comment]
This is the customer-facing issue for #1371 
A quick update - SFT LoRa is done for both megatron and PyTDtensor backend. Coming up in V0.6 is GRPO Lora and DPO Lora

***

#### DTensor V2 LoRA SFT:

LoRA is supported in the DTensor v2 pipeline via [AutoModel’s](https://docs.nvidia.com/nemo/automodel/latest/index.html) PEFT integration, with LoRA adapters injected into linear modules prior to tensor sharding and parallelization. Triton kernels are compatible with DTensor v2, though they must be disabled when tensor_parallel_size > 1 (as multi-GPU tensor parallelism does not support Triton in this context). Notably, the DTensor v1 pipeline explicitly prohibits LoRA usage.

Below is an example NeMo RL configuration:

```yaml
policy:
  dtensor_cfg:
    _v2: true
    enabled: true
    # LoRA configuration
    lora_cfg:
      enabled: true
      # Targeting options (choose one strategy)
      match_all_linear: true        # apply to all linear layers
      target_modules: []            # or list patterns, e.g. ["q_proj","k_proj","*up_proj*"]
      exclude_modules: []           # modules to exclude when not matching all
      # Capacity and scaling
      dim: 8                        # LoRA rank (r)
      alpha: 32                     # scaling; effective lr multiplier = alpha/dim
      # Regularization and init
      dropout: 0.0
      dropout_position: "post"      # "pre" or "post"
      lora_A_init: "xavier"         # "xavier" or "uniform"
      # Performance
      use_triton: true              # set to false when tensor_parallel_size > 1
```
Example recipe: [examples/configs/recipes/llm/sft-llama3.1-8b-1n8g-fsdp2tp1-lora.yaml](https://github.com/NVIDIA-NeMo/RL/blob/r0.5.0/examples/configs/recipes/llm/sft-llama3.1-8b-1n8g-fsdp2tp1-lora.yaml)

```sh
uv run python examples/run_sft.py policy.dtensor_cfg.lora_cfg.enabled=True
```

#### Megatron LoRA SFT:

Megatron LoRA SFT is enabled through [Megatron-Bridge’s](https://docs.nvidia.com/nemo/megatron-bridge/latest/training/peft.html) PEFT integration. 

Below is an example NeMo RL configuration:

```yaml
policy:
  megatron_cfg:
    peft:
      enabled: false
      target_modules: []
      exclude_modules: []
      dim: 8
      alpha: 32
      dropout: 0.0
      dropout_position: "post"
      lora_A_init_method: "xavier"
      lora_B_init_method: "zero"
      a2a_experimental: false
      lora_dtype: None
```

```sh
uv run examples/run_sft.py \
  policy.dtensor_cfg.enabled=false \
  policy.megatron_cfg.enabled=true \
  policy.megatron_cfg.lora_cfg.enabled=true
```

**Comment by qiaochuz-nv** [comment]
## QA RCCA Analysis

RCCA Status: Done
RCCA Category: covered
Test Action: none

Fix PR: https://github.com/NVIDIA-NeMo/RL/pull/1889
Root Cause: fix PR includes upstream test coverage.
Trigger Config: Support for PEFT in GPRO

Coverage Assessment:
| Area | Result |
|---|---|
| Existing Test | Upstream fix PR test coverage |
| Coverage Status | COVERED |
| Gap | N/A |
| Corrective Action | No nmfw_tests addition required; fix PR includes upstream test coverage or equivalent regression coverage. |

Conclusion: RCCA classification completed for this closed NeMo-RL GitHub issue.
